devmapper: Little code cleanup and code reorganization

rhvgoyal opened the pull request:
There are 3 patches which do following.
- Move delete transaction operation into a separate function.
-  Save deviceset metadata file little early in shutdown sequence.
- Move issue discard operation into a separate function.

rhvgoyal commented:
CC @LK4D4 @calavera @vbatts

LK4D4 commented:
@rhvgoyal You have misspell in third commit.

rhvgoyal commented:
@LK4D4 I noticed a wrong spelling in debug message and another one in commit message. Fixed it.

jfrazelle commented:
can you squash the commits please

rhvgoyal commented:
@jfrazelle I have squashed all the 3 patches into one. PTAL.

thaJeztah commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1469,40 +1469,69 @@ func (devices *DeviceSet) AddDevice(hash, baseHash string) error {
 
-	if err := devices.unregisterDevice(info.DeviceID, info.Hash); err != nil {
+	// If device open count is not zero, device is in use and skip
+	// issuing discards.
```
Nitpicking; I think this comment is a bit redundant, given that it almost literally describes the code at line 1511

rhvgoyal replied:
@thaJeztah ok, will get rid of this comment.

thaJeztah commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1469,40 +1469,69 @@ func (devices *DeviceSet) AddDevice(hash, baseHash string) error {
+		devices.issueDiscard(info)
+	}
+
+	// Try to deactivate deivce in case it is active.
```
Typo s/deivce/device/

rhvgoyal replied:
Will fix.

thaJeztah commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1657,6 +1686,9 @@ func (devices *DeviceSet) Shutdown() error {
 	var devs []*devInfo
 
 	devices.Lock()
+	// Save DeviceSet Metadata first
```
If there's a specific reason this should be done first, could you describe the *reason* here (so that other people will not move it around)? Otherwise this comment is redundant

rhvgoyal replied:
docker kills all the threads if shutdown does not finish with-in certain time (I think 10 seconds). So if some device deactivation takes a long time (as device is busy), then it is possible that docker kills the Shutdown() routine and we never get a chance to save deviceset metadata. That's why save it first before moving on to deactivating devices. 

I will put this information in comments.

thaJeztah replied:
Thanks, I think it's good to have some context

rhvgoyal commented:
@thaJeztah Fixed the issues your raised and pushed new version of patch. PTAL.

thaJeztah commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1657,6 +1684,14 @@ func (devices *DeviceSet) Shutdown() error {
+	// routine does not finish in time as we loop trying to deactivate
+	// some devices while these are busy. In that case shutdown() routine
+	// will be killed and we will not get a chance to save deviceset
+	// metadata. Hence save this early befor trying to deactivate devices.
```
typo: s/befor/before/ (no need to fix, unless there are other changes to be made)

rhvgoyal replied:
I will fix it and push again anyway. I need to employ some tools which catch these kind of typos and spelling mistakes.

thaJeztah commented:
Thanks @rhvgoyal! comments LGTM (not a code reviewer, so I'll leave that to the other maintainers)

rhvgoyal committed:
devmapper: Few code cleanups

This patch does three things. Following are the descriptions.

===
Create a separate function for delete transactions so that parent function
is little smaller.

Also close transaction if an error happens.
===
When docker is being shutdown, save deviceset metadata first before
trying to remove the devices. Generally caller gives only 10 seconds
for shutdown to complete and then kills it after that. So if some device
is busy, we will wait 20 seconds for it removal and never be able to save
metadata. So first save metadata and then deal with device removal.
===
Move issue discard operation in a separate function. This makes reading code
little easier.

Also don't issue discards if device is still open. That means devices is
still probably being used and issuing discards is not a good idea.

This is especially true in case of deferred deletion. We want to issue
discards when device is not open. At that time device can be deleted too.
Otherwise we will issue discards and deletion will actually fail. Later
we will try deletion again and issue discards again and deletion will
fail again as device is open and busy.

So this will ensure that discards are issued once when device is not open
and it can actually be deleted.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
Pushed patch one more time fixing 's/befor/before'

thaJeztah commented:
@LK4D4 @jfrazelle PTAL

vbatts commented:
LGTM

LK4D4 commented:
LGTM
